### PR TITLE
New ajax form for creating new exam template

### PR DESCRIPTION
**Pull request by jenachoi**

### CODE REVIEW

File: app/controllers/exam_templates_controller.rb [hunk 1]
```diff
@@ -9,6 +9,52 @@ def index
     @exam_templates = @assignment.exam_templates
   end
 
+  # Creates a new instance of the exam template.
+  def create
+    assignment = Assignment.find(params[:assignment_id])
+    # retrieving exam template file
```
Comment (david-yz-liu) [inline app/controllers/exam_templates_controller.rb]: This comment and the next one are unnecessary.

File: app/controllers/exam_templates_controller.rb [hunk 2]
```diff
@@ -9,6 +9,52 @@ def index
...
+    new_uploaded_io = params[:create_template][:file_io]
+    # getting filename
+    filename = params[:create_template][:filename]
```
Comment (david-yz-liu) [inline app/controllers/exam_templates_controller.rb]: Do not allow the user to input a filename on the form. The filename used should be the one of the uploaded file (from "file_io").

File: app/controllers/exam_templates_controller.rb [hunk 3]
```diff
@@ -9,6 +9,52 @@ def index
...
+    if filename.nil? || new_uploaded_io.nil? || new_uploaded_io.content_type != 'application/pdf'
+      flash_message(:error, t('exam_templates.create.failure'))
+      redirect_to action: 'index'
+      return
+    end
+    new_file_content = new_uploaded_io.read
+    # getting number of pages
+    pdf = CombinePDF.parse new_file_content
+    num_pages = pdf.pages.length
+    # if user didn't include extension of new template (which is pdf), then add '.pdf' to filename
```
Comment (david-yz-liu) [inline app/controllers/exam_templates_controller.rb]: Do not do this. See above comment.

File: app/controllers/exam_templates_controller.rb [hunk 4]
```diff
@@ -9,6 +9,52 @@ def index
...
+    if File.extname(filename) == ''
+      filename = filename + '.pdf'
+    end
+    # creating corresponding directory and file
+    assignment_name = Assignment.find(assignment.id).short_identifier
```
Comment (david-yz-liu) [inline app/controllers/exam_templates_controller.rb]: It is not necessary to call `Assignment.find` again.

File: app/views/exam_templates/_boot.js.erb [hunk 5]
```diff
@@ -0,0 +1,5 @@
+<script>
```
Comment (david-yz-liu) [inline app/views/exam_templates/_boot.js.erb]: Do not put `script` tags here. Instead, put it in the view that is rendering this partial.

File: app/views/exam_templates/_create_new_template.html.erb [hunk 6]
```diff
@@ -0,0 +1,28 @@
+<%= content_for :modal_id, 'create_new_template' %>
+<%= content_for :modal_title, t('exam_templates.create.add_new') %>
+<%= content_for :modal_content do %>
+    <%= ExamTemplate.human_attribute_name('filename') %>  <span class='required_field'>*</span>
```
Comment (david-yz-liu) [inline app/views/exam_templates/_create_new_template.html.erb]: Use 2 spaces for indentation.

File: config/locales/views/exam_templates/en.yml [hunk 7]
```diff
@@ -9,5 +9,17 @@ en:
...
+      success:
+        "Exam Template successfully created"
+      failure:
+        "Exam Template not successfully created"
+      upload:
+        "Upload Template"
+      enter_filename:
+        "Enter file name..."
 
+    save_changes: "Save Changes"
```
Comment (david-yz-liu) [inline config/locales/views/exam_templates/en.yml]: Don't create a new key for this. Reuse an existing one.

File: app/views/exam_templates/index.html.erb [hunk 8]
```diff
@@ -1,4 +1,19 @@
+<% content_for :head do %>
+    <%= render partial: 'boot',
+               formats: [:js],
```
Comment (david-yz-liu) [inline app/views/exam_templates/index.html.erb]: You don't need to specify formats or handlers.

File: app/views/exam_templates/index.html.erb [hunk 9]
```diff
@@ -1,4 +1,19 @@
...
 <% content_for :title, ExamTemplate.model_name.human.pluralize %>
 
+<% @heading_buttons = [
+    { link_text: t('exam_templates.create.add_new'),
+      link_path: '#',
+      html_options: { onclick:'modal_create_new.open()' } }
+] %>
+
+<!-- Modals -->
+<%= render partial: 'create_new_template', layout: 'layouts/modal_dialog'  %>
```
Comment (david-yz-liu) [inline app/views/exam_templates/index.html.erb]: There is a whitespace problem here.

File: app/views/exam_templates/index.html.erb (hunk 8 above)
Comment (jenachoi) [inline app/views/exam_templates/index.html.erb]: boot is taken as html if we don't specify the formats, so I only deleted handlers line!

**Comment by jenachoi** [comment]
still working on it- will finish if after Steve's meeting! : )

**Comment by jenachoi** [comment]
I almost forgot to move code in ExamTemplatesController#create into a method in the ExamTemplate model class. Oops sorry about that. I will make a change soon-

**Comment by david-yz-liu** [comment]
Note: would be good to switch over to using responders now.

**Pull request merged**